src/backend/distributed/planner/multi_router_planner.c

```diff
@@ -507,9 +562,22 @@ ModifyQuerySupported(Query *queryTree, bool multiShardQuery)
 	List *onConflictSet = NIL;
 	Node *arbiterWhere = NULL;
 	Node *onConflictWhere = NULL;
-
 	CmdType commandType = queryTree->commandType;
 
+	if (multiShardQuery)
```

Comment:
I think there was a flaw in my reasoning here. This problem is not specific to multi-shard queries, but to intermediate results which could cause you to get a CTID from some other node / transaction. 

so I think this should be: `ContainsReadIntermediateResultFunction(queryTree)`.

That way, subqueries on single shard with ctid are still allowed, provided they don't also have intermediate results.